File: labs/manipulacao/src/ts/shell.ts
const input = () => ""; // MODO_TESTE
export {};

function get_men(vet: number[]): number[] {
    // @DROP
    let output: number[] = [];
    for (let elem of vet)
        if (elem > 0)
            output.push(elem);
    return output;
    // @UNC
    // return [];
}

function get_calm_women(vet: number[]): number[] {
    // @DROP
    return vet.filter(x => x < 0 && Math.abs(x) < 10);
    // @UNC
    // return [];
}

function sort(vet: number[]): number[] {
    // @DROP
    let aux = vet.slice(); //faz uma copia
    aux.sort((a, b) => a - b);
    return aux;
    // @UNC
    // return [];
}

function sort_stress(vet: number[]): number[] {
    // @DROP
    vet.sort((a, b) => Math.abs(a) - Math.abs(b));
    return vet;
    // @UNC
    // return [];
}

function reverse(vet: number[]): number[] {
    // @DROP
    let output: number[] = [];
    for (let elem of vet)
        output.unshift(elem);
    return output;
    // @UNC
    // return [];
}

function unique(vet: number[]): number[] {
    // @DROP
    let unique_list: number[] = [];
    for (let elem of vet)
        if (!unique_list.includes(elem))
            unique_list.push(elem);
    return unique_list;
    // @UNC
    // return [];
}

function repeated(vet: number[]): number[] {
    // @DROP
    let unique_list: number[] = [];
    let repeated: number[] = [];
    for (let elem of vet) {
        if (!unique_list.includes(elem))
            unique_list.push(elem);
        else
            repeated.push(elem);
    }
    return repeated;
    // @UNC
    // return [];
}

// @KEEP

function main() {
    let vet = new Array<string>();

    while (true) {
        let line = input();
        console.log("$" + line);
        let args = line.split(" ");

        if      (args[0] === "end")             { break;                                             }
        else if (args[0] === "get_men")         { console.log(fmt(get_men(       to_vet(args[1])))); }
        else if (args[0] === "get_calm_women")  { console.log(fmt(get_calm_women(to_vet(args[1])))); }
        else if (args[0] === "sort")            { console.log(fmt(sort(          to_vet(args[1])))); }
        else if (args[0] === "sort_stress")     { console.log(fmt(sort_stress(   to_vet(args[1])))); }
        else if (args[0] === "reverse")         {
            let vet = to_vet(args[1]);
            let reversed = reverse(vet);
            console.log(fmt(vet));
            console.log(fmt(reversed));
        }
        else if (args[0] === "unique")          { console.log(fmt(unique(        to_vet(args[1])))); }
        else if (args[0] === "repeated")        { console.log(fmt(repeated(      to_vet(args[1])))); }
        else                                    { console.log("fail: comando invalido");             }
    }
}

// Função auxiliar para converter de string para vetor
// "[1,2,3,4]" para [1, 2, 3, 4]
function to_vet(token: string): number[] {
    let size = token.length;
    let inside = token.substring(1, size - 1);
    return inside === "" ? [] : inside.split(",").map(x => +x)
}

//Converte de vetor para string sem inserir os espaços
//[1, 2, 3, 4] => "[1,2,3,4]"
function fmt(vet: number[]): string {
    if (vet == null) {
        return "[]";
    }
    return "[" + vet.join(", ") + "]";
}

main()

File: labs/manipulacao/src/cpp/shell.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <sstream>

std::vector<int> get_men(std::vector<int> vet) {
    // @DROP
    std::vector<int> unique;
    for (auto elem : vet)
        if (elem > 0)
            unique.push_back(elem);
    return unique;
    // @UNC
    // (void) vet;
    // return {};
}

std::vector<int> get_calm_women(std::vector<int> vet) {
    // @DROP
    std::vector<int> unique;
    for (auto elem : vet)
        if (elem < 0 && abs(elem) < 10)
            unique.push_back(elem);
    return unique;
    // @UNC
    // (void) vet;
    // return {};
}

std::vector<int> sort(std::vector<int> vet) {
    // @DROP
    std::sort(vet.begin(), vet.end());
    return vet;
    // @UNC
    // (void) vet;
    // return {};
}

std::vector<int> sort_stress(std::vector<int> vet) {
    // @DROP
    std::sort(vet.begin(), vet.end(), [] (int a, int b) {
        return abs(a) < abs(b);
    });
    return vet;
    // @UNC
    // (void) vet;
    // return {};
}

std::vector<int> reverse(std::vector<int> vet) {
    // @DROP
    std::reverse(vet.begin(), vet.end());
    return vet;
    // @UNC
    // (void) vet;
    // return {};
}

// @DROP
auto exists = [](auto vet, auto value) {
    return std::find(vet.begin(), vet.end(), value) != vet.end();
};
// @KEEP
std::vector<int> unique(std::vector<int> vet) {
    // @DROP
    std::vector<int> unique;
    for (auto elem : vet)
        if (!exists(unique, elem))
            unique.push_back(elem);
    return unique;
    // @UNC
    // (void) vet;
    // return {};
}

std::vector<int> repeated(std::vector<int> vet) {
    // @DROP
    std::vector<int> unique;
    std::vector<int> repeated;

    for (auto elem : vet) {
        if (!exists(unique, elem))
            unique.push_back(elem);
        else
            repeated.push_back(elem);
    }
    return repeated;
    // @UNC
    // (void) vet;
    // return {};
}

std::string input() {
    std::string line;
    std::getline(std::cin, line);
    return line;
}

template <typename T>
void print(T t) {
    std::cout << t << std::endl;
}

void print(std::vector<int> vet) {
    std::cout << '[';
    for (int i = 0; i < (int) vet.size(); i++) {
        std::cout << (i == 0 ? "" : ", ");
        std::cout << vet[i];
    }
    std::cout << ']' << std::endl;
}

std::vector<std::string> split(const std::string &s, char delim) {
    std::vector<std::string> elems;
    std::string elem;
    for (auto c : s) {
        if (c == delim) {
            elems.push_back(elem);
            elem.clear();
        } else {
            elem += c;
        }
    }
    elems.push_back(elem);
    return elems;
}

std::vector<int> str2vet(std::string s) {
    s = s.substr(1, s.size() - 2);
    std::vector<int> vet;
    std::stringstream ss(s);
    std::string elem;
    while (std::getline(ss, elem, ',')) {
        vet.push_back(std::stoi(elem));
    }
    return vet;
}

int main() {
    
    while(true) {
        std::string line = input();
        auto args = split(line, ' ');
        print('$' + line);

        if     (args[0] == "get_men"        ) { print(get_men(str2vet(args[1])));        }
        else if(args[0] == "get_calm_women" ) { print(get_calm_women(str2vet(args[1]))); }
        else if(args[0] == "sort"           ) { print(sort(str2vet(args[1])));           }
        else if(args[0] == "sort_stress"    ) { print(sort_stress(str2vet(args[1])));    }
        else if(args[0] == "reverse"        ) {
            auto array = str2vet(args[1]);
            auto other = reverse(array);
            print(array);
            print(other);
        }
        else if(args[0] == "unique"         ) { print(unique(str2vet(args[1])));         }
        else if(args[0] == "repeated"       ) { print(repeated(str2vet(args[1])));       }
        else if(args[0] == "end"            ) { break;                                   }
    }
}
